### PR TITLE
feat: av pro implementation

### PR DESCRIPTION
**Pull request by dalkia**
## What does this PR change?

Add the AVPro free plugin to be used as a video streaming framework.

## How to test the changes?

1) Open the desktop version using https://renderer-artifacts.decentraland.org/desktop/index.html?dcl://DESKTOP-BRANCH=feat/av-pro-implementation
2) Go to the following video links, and check that they are working correctly, both audio and video. All of the videos should have the RENDERHEADS watermark since we are using the free version of AVPro.
3) Test both in Windows and in Mac

- Filmrare (2,55)
Go to Root.AX. There are rooms with a video that shows different authors. No audio here
- Notus Finanse. (-22,37)
There is a CEO speaking just next to you
- DG Arena. (-118,120)
Go into the arena. There are different screens in the corners, as well as a poker table in the middle
- Wilderness. (-24,89)
There is a big screen with a tutorial on a corner of the scene. Look for a youtube icon, on top of that
- Art week. 
There are many experiences using video. Have a look around!
55,61
58,60
62,61
66,62
61,66
54,64
61,70
53,73
64,70
69,65
55,65 (Metaboralis, causes plenty of issues due to a large amount of videos. There is some problem handling in this scene)

## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md

### CODE REVIEW

File: unity-renderer-desktop/Assets/Scripts/MainScripts/DCL/MainDesktop/MainDesktop.cs [hunk 1]
```diff
@@ -21,13 +22,22 @@ public class MainDesktop : Main
...
+        private BaseVariable<FeatureFlag> featureFlags => DataStore.i.featureFlags.flags;
 
         protected override void Awake()
         {
             CommandLineParserUtils.ParseArguments();
             isConnectionLost = false;
-
-            DCLVideoTexture.videoPluginWrapperBuilder = () => new VideoPluginWrapper_Native();
+           
+            if (featureFlags.Get().IsFeatureEnabled("use_avpro_player") && Application.platform != RuntimePlatform.LinuxPlayer)
```
Comment (Antivortex) [inline unity-renderer-desktop/Assets/Scripts/MainScripts/DCL/MainDesktop/MainDesktop.cs:32]: Maybe these combined condition could be embedded into some AVPro wrapper class or even polymorphic video player class with a factory

**Review by Antivortex — APPROVED** [review]

**Review by kwspgoliasz — CHANGES_REQUESTED** [review]
The feature appears to be implemented correctly on the Windows App.
Both the videos and audio are correctly displayed, however, in some cases it can take a while for the video to load (ex. Art Week's Metaboralis area).

On the Mac version of the app, small images with motion tend to be displayed correctly, however, in case of larger movies (especially on big screens), the videos either do not load, or get permanently frozen. This is especially visible in the DG Arena scene.

**Comment by kwspgoliasz** [comment]
The fixes for the Mac resulted in noticeable improvements. The issue with the permanent video freeze is no longer present, however, short freezes for the more demanding videos can still be noticed. When the issue occurs, the video is always resumed after the freeze which lasts for about 10 sec.. This issue is most prominent in the Makersplace (55,62) and at the DG Arena (-118,120) for the corner screens.  
Additionally, an issue with the video not loading at all occurred at the Art Week Main Building area (62,61) and for some objects in the Meta Borealis (55,65).

**Comment by dalkia** [comment]
Art week building video is not supported by the mac codec

File: unity-renderer-desktop/Assets/Scripts/MainScripts/DCL/Components/Video/AvPro/AvProVideoPlayer.cs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    {
+        avProMediaPlayer.PlaybackRate = playbackRate;
+    }
+
+    public void SetLoop(bool doLoop)
+    {
+        this.doLoop = doLoop;
+    }
+
+    private IEnumerator ResizeVideoTexture()
```
Comment (AjimenezDCL) [inline unity-renderer-desktop/Assets/Scripts/MainScripts/DCL/Components/Video/AvPro/AvProVideoPlayer.cs:142]: We can use a task here

**Review by AjimenezDCL — CHANGES_REQUESTED** [review]
Great job, super clean. I just left some smal requestes

**Review by kwspgoliasz — CHANGES_REQUESTED** [review]
Most of the videos are properly displayed on both, the MacOS and Windows clients, however, some issues are still present.

The issue with some videos not loading can be observed in the following locations:

- DG Arena - screens at the center of the hall and on specific floors - https://user-images.githubusercontent.com/110464313/196973942-5a8e36a8-ac48-4c2f-920e-a8569ae44fd2.mp4.

- Wilderness - the tutorial video - https://user-images.githubusercontent.com/110464313/196974130-0528243e-bde5-4ae3-8fc3-f88cb1c0990c.mp4.

- Filmrare - the cinema screen - https://user-images.githubusercontent.com/110464313/196973975-dd925fa3-857c-4516-ae68-e95af11d98e4.mp4.

Moreover, many of the videos on MacOS suffer from white screen flashing. On the Windows client, the issue occurs only for the VVSVS scene - https://user-images.githubusercontent.com/110464313/196974221-b4c68d5e-e316-467d-bdee-c917a1230455.mp4. 
Slight freezing can be observed when watching the videos on the Mac, however, this might be caused by MacBook's low-end hardware.

Additionally, some of the videos may sometimes require a longer loading time. This issue was observed exclusively in the Art Week area - https://user-images.githubusercontent.com/110464313/196975174-8e9567f0-63e7-451a-96b0-b58ee16f65ec.mp4.

**Comment by dalkia** [comment]
> Most of the videos are properly displayed on both, the MacOS and Windows clients, however, some issues are still present.
> 
> The issue with some videos not loading can be observed in the following locations:
> 
> * DG Arena - screens at the center of the hall and on specific floors - https://user-images.githubusercontent.com/110464313/196973942-5a8e36a8-ac48-4c2f-920e-a8569ae44fd2.mp4.
> * Wilderness - the tutorial video - https://user-images.githubusercontent.com/110464313/196974130-0528243e-bde5-4ae3-8fc3-f88cb1c0990c.mp4.
> * Filmrare - the cinema screen - https://user-images.githubusercontent.com/110464313/196973975-dd925fa3-857c-4516-ae68-e95af11d98e4.mp4.
> 
> Moreover, many of the videos on MacOS suffer from white screen flashing. On the Windows client, the issue occurs only for the VVSVS scene - https://user-images.githubusercontent.com/110464313/196974221-b4c68d5e-e316-467d-bdee-c917a1230455.mp4. Slight freezing can be observed when watching the videos on the Mac, however, this might be caused by MacBook's low-end hardware.
> 
> Additionally, some of the videos may sometimes require a longer loading time. This issue was observed exclusively in the Art Week area - https://user-images.githubusercontent.com/110464313/196975174-8e9567f0-63e7-451a-96b0-b58ee16f65ec.mp4.

The three videos mentioned (Wilderness, Center video in DG Arena and Filmrare cinema) have been taken down. They are not working either in prd in webgl.

File: unity-renderer-desktop/Assets/Scripts/MainScripts/DCL/Components/Video/AvPro/AvProVideoPlayer.cs [hunk 3]
```diff
@@ -13,9 +14,11 @@ public class AvProVideoPlayer : IVideoPlayer, IDisposable
     private Texture avProTexture;
     private string lastError;
     private string id;
-    private bool resizeDone;
     private bool doLoop;
 
+    private CancellationTokenSource resizeVideoTextureCancellationSource;
```
Comment (AjimenezDCL) [inline unity-renderer-desktop/Assets/Scripts/MainScripts/DCL/Components/Video/AvPro/AvProVideoPlayer.cs:19]:
```suggestion
    private CancellationTokenSource resizeVideoTextureCts;
```

the `Cts` acronym is widely accepted for CancellationTokenSource. I leave it up to you if you prefer the long name.

**Review by AjimenezDCL — APPROVED** [review]

**Pull request merged**